packages/hawtio/src/plugins/camel/routes/Source.tsx

```diff
@@ -1,24 +1,76 @@
 import { CamelContext } from '@hawtiosrc/plugins/camel/context'
-import { CodeEditor, Language } from '@patternfly/react-code-editor'
+import { CodeEditor, CodeEditorControl, Language } from '@patternfly/react-code-editor'
 import React, { useContext, useEffect, useState } from 'react'
 import { log } from '../globals'
+import { CheckCircleIcon, SaveIcon } from '@patternfly/react-icons'
+import { isRouteNode, isRoutesFolder } from '@hawtiosrc/plugins/camel/camel-service'
+import { routesService } from './routes-service'
+import { eventService } from '@hawtiosrc/core'
 
 export const Source: React.FunctionComponent = () => {
   const { selectedNode } = useContext(CamelContext)
   const [xmlSource, setXmlSource] = useState('')
+  const [isUpdateEnabled, setIsUpdateEnabled] = useState(false)
+  const [codeChanged, setCodeChanged] = useState(false)
+  const isRoute: boolean = isRouteNode(selectedNode!) && !isRoutesFolder(selectedNode!)
 
   useEffect(() => {
-    const xml = selectedNode?.getMetadata('xml')
-    if (xml) {
-      setXmlSource(xml)
-    } else {
-      log.warn('Source - Unable to fetch XML from', selectedNode)
+    if (selectedNode) {
+      const xml = selectedNode?.getMetadata('xml')
+
+      if (isRoute) {
+        routesService.isRouteUpdateEnabled(selectedNode!).then(enabled => {
+          setIsUpdateEnabled(enabled)
+        })
+      }
+
+      if (xml) {
+        setXmlSource(xml)
+      } else {
+        log.warn('Source - Unable to fetch XML from', selectedNode)
+      }
     }
-  }, [selectedNode])
+  }, [isRoute, selectedNode])
+
+  const onCodeChange = (code: string) => {
+    setCodeChanged(code !== xmlSource)
+  }
+
+  const onSaveClick = (code: string) => {
+    if (isRoute && code !== xmlSource) {
+      try {
+        routesService.saveRoute(selectedNode!, code)
```

Comment:
same here: `!`

If we need `selectedNode` to exist, check it's existence right after the `useEffect` and then you are free from null-checking of it afterwards. See how the other components handle that.